tests/test_proto_methods.rs

```diff
@@ -82,6 +82,7 @@ fn test_getattr() {
         let example_py = make_example(py);
         assert_eq!(
             example_py
+                .as_any()
```

Comment:
These `.as_any()` calls are necessary, because the `Deref` implementation didn't work? Might be worth checking that, I wonder if we need to adjust.

Reply:
Yeah, we get errors like:

```rust
error[E0599]: no method named `getattr` found for struct `pyo3::Bound` in the current scope
  --> tests/test_proto_methods.rs:85:18
   |
83 | /             example_py
84 | |                 //.as_any()
85 | |                 .getattr("value")
   | |                 -^^^^^^^ method not found in `Bound<'_, ExampleClass>`
   | |_________________|
   | 

For more information about this error, try `rustc --explain E0599`.
```

Reply:
Hmm. I think probably we need to adjust the bounds of the `Deref` implementation so that it works in this case, I've added that to the tracking issue. If you have an idea, feel free to open a PR. Otherwise, I'll sleep on it and see what ideas I have tomorrow.

Reply:
I'm wondering if we can solve this by implementing `AsRef<PyAny>` in the `#[pyclass]` proc macro. That would make the existing `Deref` bounds sufficient.

Reply:
I have noticed this too recently. I believe this happens only if `T` in `Bound<T>` is a `#[pyclass]` type. I don't think implementing `AsRef<PyAny>` is the right path here. I believe the current bound is wrong, because we're not using the `AsRef` functionality in any way. It just happens to work (mostly) because we're implementing that for the native types, but it feels like we're misusing it as a marker.

I believe an explicit `AsRef` might still be usefull, but it would need to be `AsRef<Bound<PyAny>>`, so I think the `Deref` bound could already be `Bound<'py, T>: AsRef<Bound<'py, PyAny>>,`.

Then we would need to find a way to correct the `impl<'py, T> AsRef<Bound<'py, PyAny>> for Bound<'py, T>` blanket implementation, because it uses the same hack. This can sadly not be implemented by the macro code because of orphan rules, so maybe using a marker trait together with a blanket implementation is a path that we can take.

Reply:
Yes, I think that's the right solution too. I haven't thought of a nice form for the marker trait, presumably it would need to be implemented for basically all `T: PyTypeInfo` _except_ for `PyAny`.

Reply:
A possibly nice idea for the marker trait is something like `HasPyBaseType` (as `PyAny` doesn't), but that might imply slightly more complex structure than what we really need here.